File: packages/agentspec/agentspec/prompts.py
"""Prompt templates for agent pipeline operations.

Centralised, parameterizable prompts for PRD assessment, refinement, and
artifact generation.  Each function constructs the system or user message
sent to the Anthropic messages API.

Prompt content is loaded from markdown template files under
``_templates/prompts/``, with project-level overrides in
``.spec/prompts/`` taking precedence.
"""

from __future__ import annotations

import json
from pathlib import Path
from typing import TYPE_CHECKING, Any

from agentspec.prompt_loader import load_prompt, load_prompt_template

if TYPE_CHECKING:
    from agentspec.session import Assessment

# ── helpers ──────────────────────────────────────────────────────────


def _require_non_empty(value: str, name: str) -> None:
    """Raise ``ValueError`` if *value* is empty or whitespace-only."""
    if not value or not value.strip():
        raise ValueError(f"{name} must not be empty")


def _format_assessment_block(previous_assessment: Assessment) -> str:
    """Format an assessment's quality, summary, and gaps into a text block."""
    block = f"Quality: {previous_assessment.quality}\nSummary: {previous_assessment.summary}\n"
    if previous_assessment.gaps:
        block += "Gaps:\n"
        for gap in previous_assessment.gaps:
            block += f"  - {gap}\n"
    return block


def _format_qa_block(
    previous_assessment: Assessment,
    answers: dict[str, str],
) -> str:
    """Format questions and answers into a text block."""
    parts: list[str] = []
    for q in previous_assessment.questions:
        answer_text = answers.get(q.id, "(no answer provided)")
        parts.append(
            f"- {q.id}: {q.text}\n  Context: {q.context}\n  Answer: {answer_text}"
        )
    return "\n".join(parts)


def _format_prior_artifacts(prior_artifacts: dict[str, Any] | None) -> str:
    """Format previously generated artifacts as a context section."""
    if not prior_artifacts:
        return ""
    parts = ["## Previously Generated Artifacts\n"]
    for name, content in prior_artifacts.items():
        parts.append(f"### {name}\n\n```json\n{json.dumps(content, indent=2)}\n```\n")
    return "\n".join(parts)


def _format_spec_landscape(landscape: list[dict[str, Any]] | None) -> str:
    """Format a landscape list into a markdown section for LLM prompt injection.

    Returns an empty string when *landscape* is ``None`` or empty, ensuring
    backward-compatible prompt output.  Otherwise builds a
    ``## Existing Spec Landscape`` section with tables for active and/or
    archived specs.
    """
    if not landscape:
        return ""

    active = [e for e in landscape if not e.get("archived", False)]
    archived = [e for e in landscape if e.get("archived", False)]

    parts: list[str] = [
        "## Existing Spec Landscape\n",
        "The following specs already exist in this project. "
        "Check for overlaps, historical precedent, and potential dependencies.\n",
    ]

    if active:
        parts.append("### Active Specs\n")
        parts.append("| Spec | Title | Status | Intent |")
        parts.append("|------|-------|--------|--------|")
        for entry in active:
            spec = entry.get("spec_name", entry.get("spec_id", ""))
            title = entry.get("title", "")
            status = entry.get("status", "")
            intent = entry.get("intent", "")
            glossary_terms = entry.get("glossary_terms", [])
            if glossary_terms:
                terms_str = ", ".join(glossary_terms[:10])
                intent = (
                    f"{intent} (Terms: {terms_str})"
                    if intent
                    else f"Terms: {terms_str}"
                )
            parts.append(f"| {spec} | {title} | {status} | {intent} |")
        parts.append("")

    if archived:
        parts.append("### Archived Specs\n")
        parts.append("| Spec | Title | Status |")
        parts.append("|------|-------|--------|")
        for entry in archived:
            spec = entry.get("spec_name", entry.get("spec_id", ""))
            title = entry.get("title", "")
            status = entry.get("status", "")
            parts.append(f"| {spec} | {title} | {status} |")
        parts.append("")

    return "\n".join(parts)


def _format_dependent_interfaces(
    dependent_interfaces: list[dict[str, Any]] | None,
) -> str:
    """Format dependent spec interfaces into a markdown section."""
    if not dependent_interfaces:
        return ""

    parts: list[str] = [
        "## Dependent Spec Interfaces\n",
        "The following interfaces are defined by upstream specs that this spec depends on. "
        "Use the exact names, types, and signatures below.\n",
    ]

    for iface in dependent_interfaces:
        parts.append(f"### Spec {iface['spec_id']} — {iface.get('spec_name', '')}\n")

        glossary = iface.get("glossary", {})
        if glossary:
            parts.append("**Glossary:**\n")
            for term, definition in glossary.items():
                parts.append(f"- **{term}**: {definition}")
            parts.append("")

        external_apis = iface.get("external_apis", [])
        if external_apis:
            parts.append("**External APIs:**\n")
            parts.append(f"```json\n{json.dumps(external_apis, indent=2)}\n```\n")

        symbols = iface.get("interface_symbols", [])
        if symbols:
            parts.append("**Interface Symbols:**\n")
            parts.append("| Criterion ID | Action | Return Contract |")
            parts.append("|---|---|---|")
            for sym in symbols:
                cid = sym.get("criterion_id", "")
                action = sym.get("action", "").replace("\n", " ")
                rc = sym.get("return_contract", "").replace("\n", " ")
                parts.append(f"| {cid} | {action} | {rc} |")
            parts.append("")

    return "\n".join(parts)


_LANGUAGE_MARKERS: list[tuple[str, str, str]] = [
    ("go.mod", "Go", "go test ./... -count=1, go vet ./..."),
    ("Cargo.toml", "Rust", "cargo test, cargo clippy"),
    ("package.json", "TypeScript/JavaScript", "npm test, eslint ."),
    ("pyproject.toml", "Python", "pytest, ruff check"),
    ("setup.py", "Python", "pytest, ruff check"),
    ("build.gradle", "Java/Kotlin", "gradle test, gradle check"),
    ("pom.xml", "Java", "mvn test, mvn checkstyle:check"),
    ("mix.exs", "Elixir", "mix test, mix credo"),
    ("Gemfile", "Ruby", "bundle exec rspec, rubocop"),
]


def _detect_project_language(project_dir: Path | None) -> tuple[str, str] | None:
    """Detect the project's primary language from manifest files.

    Returns a ``(language, tooling_hint)`` tuple, or ``None`` if no
    marker file is found.
    """
    if not project_dir:
        return None
    for filename, language, tooling in _LANGUAGE_MARKERS:
        if (project_dir / filename).exists():
            return language, tooling
    return None


def _format_project_context(project_dir: Path | None) -> str:
    """Build a ``## Project Context`` block from detected language."""
    detected = _detect_project_language(project_dir)
    if not detected:
        return ""
    language, tooling = detected
    return (
        f"## Project Context\n\n"
        f"This is a **{language}** project. All test commands, verification "
        f"checks, file paths, code patterns, and stub/dead-code markers MUST "
        f"use {language} conventions (e.g. {tooling}).\n"
    )


# ── assessment ───────────────────────────────────────────────────────


def assessment_system_prompt(*, project_dir: Path | None = None) -> str:
    """Return the system prompt for PRD assessment.

    Instructs the model to evaluate PRD quality against spec-format
    expectations, explicitly checking for the Intent, Goals, Non-Goals,
    and Background sections.
    """
    return load_prompt("assessment_system", project_dir=project_dir)


def assessment_user_prompt(
    prd_text: str,
    spec_name: str,
    *,
    project_dir: Path | None = None,
    spec_landscape: list[dict[str, Any]] | None = None,
) -> str:
    """Return the user message for PRD assessment.

    When *spec_landscape* is provided, the formatted landscape markdown
    is injected into the prompt via ``$spec_landscape_block``.

    Raises ``ValueError`` if *prd_text* is empty.
    """
    _require_non_empty(prd_text, "prd_text")

    spec_landscape_block = _format_spec_landscape(spec_landscape)

    return load_prompt_template(
        "assessment_user",
        project_dir=project_dir,
        prd_text=prd_text,
        spec_name=spec_name,
        spec_landscape_block=spec_landscape_block,
    )


# ── refinement ───────────────────────────────────────────────────────


def refinement_system_prompt(*, project_dir: Path | None = None) -> str:
    """Return the system prompt for PRD refinement.

    Instructs the model to incorporate the user's answers into the PRD
    and re-assess the updated document.
    """
    return load_prompt("refinement_system", project_dir=project_dir)


def refinement_user_prompt(
    prd_text: str,
    answers: dict[str, str],
    previous_assessment: Assessment,
    *,
    project_dir: Path | None = None,
    spec_landscape: list[dict[str, Any]] | None = None,
) -> str:
    """Return the user message for PRD refinement.

    Formats the original PRD, the user's answers (keyed by question ID),
    and the previous assessment into a single user message.  When
    *spec_landscape* is provided, the formatted landscape markdown is
    injected via ``$spec_landscape_block``.
    """
    assessment_block = _format_assessment_block(previous_assessment)
    qa_block = _format_qa_block(previous_assessment, answers)
    spec_landscape_block = _format_spec_landscape(spec_landscape)

    return load_prompt_template(
        "refinement_user",
        project_dir=project_dir,
        prd_text=prd_text,
        assessment_block=assessment_block,
        qa_block=qa_block,
        spec_landscape_block=spec_landscape_block,
    )


# ── generation ───────────────────────────────────────────────────────


def generation_system_prompt(*, project_dir: Path | None = None) -> str:
    """Return the system prompt for artifact generation.

    Instructs the model to produce a single artifact at a time in the
    correct JSON schema, conforming to spec-format v1.2.
    """
    return load_prompt("generation_system", project_dir=project_dir)


def generation_user_prompt(
    prd_text: str,
    artifact_name: str,
    prior_artifacts: dict[str, Any] | None = None,
    *,
    spec_id: str = "",
    project_dir: Path | None = None,
    dependent_interfaces: list[dict[str, Any]] | None = None,
    spec_landscape: list[dict[str, Any]] | None = None,
) -> str:
    """Return the user message for generating one artifact.

    *prior_artifacts* is a dict of already-generated artifacts
    (e.g., ``{"requirements": {...}}``) to provide as context.
    *spec_id* is the spec identifier used as prefix in all IDs.
    *dependent_interfaces* is a list of interface summaries from
    upstream dependency specs.
    *spec_landscape* is an optional list of existing spec metadata
    for cross-spec awareness during generation.

    Raises ``ValueError`` if *prd_text* is empty.
    """
    _require_non_empty(prd_text, "prd_text")

    spec_id_block = ""
    if spec_id:
        spec_id_block = (
            f"The spec_id for this spec is `{spec_id}`. Use it as the "
            f"prefix in all IDs (e.g. `{spec_id}-REQ-1`, "
            f"`TS-{spec_id}-1`).\n"
        )

    prior_artifacts_block = _format_prior_artifacts(prior_artifacts)
    project_context_block = _format_project_context(project_dir)
    dependent_interfaces_block = _format_dependent_interfaces(dependent_interfaces)
    spec_landscape_block = _format_spec_landscape(spec_landscape)

    additional_instructions = ""
    try:
        additional_instructions = load_prompt(
            f"generation_user_{artifact_name}",
            project_dir=project_dir,
        )
    except (FileNotFoundError, ValueError):
        pass

    return load_prompt_template(
        "generation_user_base",
        project_dir=project_dir,
        artifact_name=artifact_name,
        spec_id_block=spec_id_block,
        project_context_block=project_context_block,
        prd_text=prd_text,
        dependent_interfaces_block=dependent_interfaces_block,
        spec_landscape_block=spec_landscape_block,
        prior_artifacts_block=prior_artifacts_block,
        additional_instructions=additional_instructions,
    )


# ── repair ──────────────────────────────────────────────────────────


def repair_user_prompt(
    artifact_name: str,
    original_content: dict[str, Any],
    errors: list[str],
    *,
    project_dir: Path | None = None,
) -> str:
    """Return a user message asking the LLM to fix validation errors.

    Sends the original artifact content and a list of errors, asking
    the model to resubmit with corrections.
    """
    error_list = "\n".join(f"- {e}" for e in errors)
    original_json = json.dumps(original_content, indent=2)

    return load_prompt_template(
        "repair_user",
        project_dir=project_dir,
        artifact_name=artifact_name,
        error_list=error_list,
        original_json=original_json,
    )

File: golang/agentspec/prompts.go
package agentspec

import (
	"embed"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
)

// templateFS embeds all prompt template files at compile time.
//
//go:embed templates/*.md
var templateFS embed.FS

// PromptTemplateNames lists the 10 embedded prompt template names.
var PromptTemplateNames = []string{
	"assessment_system",
	"assessment_user",
	"refinement_system",
	"refinement_user",
	"generation_system",
	"generation_user_base",
	"generation_user_requirements",
	"generation_user_test_spec",
	"generation_user_tasks",
	"repair_user",
}

// validNameRe matches safe template names: [a-zA-Z0-9_-]+
var validNameRe = regexp.MustCompile(`^[a-zA-Z0-9_-]+$`)

// stripFrontmatter removes YAML frontmatter (content between opening and
// closing --- delimiters at the start of a file) from the given text.
// If no frontmatter is present, the text is returned unchanged.
func stripFrontmatter(text string) string {
	if !strings.HasPrefix(text, "---") {
		return text
	}
	// Find the closing --- delimiter after the opening one.
	rest := text[3:]
	idx := strings.Index(rest, "\n---")
	if idx < 0 {
		return text
	}
	// Skip past the closing --- and any trailing newline.
	after := rest[idx+4:]
	if strings.HasPrefix(after, "\n") {
		after = after[1:]
	}
	return after
}

// LoadPrompt loads a named prompt template from a project-local override path
// or the embedded defaults, stripping YAML frontmatter. The name must match
// [a-zA-Z0-9_-]+ to prevent path traversal.
//
// When a non-symlink override file exists at <projectDir>/.spec/prompts/<name>.md,
// it is used. Otherwise, the embedded default is used.
//
// Returns (content string, nil) on success, or ("", error) on failure.
func LoadPrompt(name, projectDir string) (string, error) {
	// Validate name to prevent path traversal.
	if !validNameRe.MatchString(name) {
		return "", fmt.Errorf("LoadPrompt: invalid template name %q", name)
	}

	// Try project-local override if projectDir is specified.
	if projectDir != "" {
		overridePath := filepath.Join(projectDir, ".spec", "prompts", name+".md")
		info, err := os.Lstat(overridePath)
		if err == nil && !info.IsDir() {
			// Reject symlinks — fall back to embedded default.
			if info.Mode()&os.ModeSymlink != 0 {
				// Symlink detected, ignore override.
			} else {
				data, err := os.ReadFile(overridePath)
				if err != nil {
					return "", fmt.Errorf("LoadPrompt: reading override file: %w", err)
				}
				return stripFrontmatter(string(data)), nil
			}
		}
	}

	// Fall back to embedded default.
	data, err := templateFS.ReadFile("templates/" + name + ".md")
	if err != nil {
		return "", fmt.Errorf("LoadPrompt: reading embedded template %q: %w", name, err)
	}
	return stripFrontmatter(string(data)), nil
}

// LoadPromptTemplate loads a prompt template via LoadPrompt and performs safe
// $variable substitution. Unmatched $variable references pass through unchanged.
//
// Returns (substitutedContent string, nil) on success, or ("", error) on failure.
func LoadPromptTemplate(name, projectDir string, vars map[string]string) (string, error) {
	content, err := LoadPrompt(name, projectDir)
	if err != nil {
		return "", err
	}

	// Perform safe substitution: replace $varname with value from vars.
	// Unmatched variables pass through unchanged.
	for k, v := range vars {
		content = strings.ReplaceAll(content, "$"+k, v)
	}

	return content, nil
}

// manifestEntry describes a project manifest file and the language/tooling it maps to.
type manifestEntry struct {
	filename  string
	language  string
	toolHints string
}

// manifestOrder defines the priority order for manifest file detection.
var manifestOrder = []manifestEntry{
	{"go.mod", "Go", "Go modules (go mod), go test, go build, go vet"},
	{"Cargo.toml", "Rust", "Cargo (cargo build, cargo test, cargo clippy)"},
	{"package.json", "TypeScript", "npm/yarn/pnpm, TypeScript compiler (tsc), Jest/Vitest"},
	{"pyproject.toml", "Python", "pip/uv/poetry, pytest, mypy/pyright, ruff"},
	{"Gemfile", "Ruby", "Bundler (bundle install), RSpec, RuboCop"},
	{"build.gradle", "Java", "Gradle (gradle build, gradle test)"},
	{"pom.xml", "Java", "Maven (mvn compile, mvn test)"},
}

// DetectProjectLanguage scans a project directory for manifest files and returns
// the detected language name and tooling hints string. Returns ("", "") when no
// recognized manifest files are present.
func DetectProjectLanguage(projectDir string) (language string, toolHints string) {
	for _, m := range manifestOrder {
		path := filepath.Join(projectDir, m.filename)
		if _, err := os.Stat(path); err == nil {
			return m.language, m.toolHints
		}
	}
	return "", ""
}

// FormatSpecLandscape formats spec landscape entries into markdown tables
// split into active and archived sections. Returns an empty string for an
// empty input slice.
func FormatSpecLandscape(landscape []map[string]any) string {
	if len(landscape) == 0 {
		return ""
	}

	var active, archived []map[string]any
	for _, entry := range landscape {
		status, _ := entry["status"].(string)
		if strings.EqualFold(status, "archived") {
			archived = append(archived, entry)
		} else {
			active = append(active, entry)
		}
	}

	var sb strings.Builder

	writeTable := func(entries []map[string]any) {
		sb.WriteString("| Spec ID | Title | Status |\n")
		sb.WriteString("|---------|-------|--------|\n")
		for _, e := range entries {
			specID, _ := e["spec_id"].(string)
			title, _ := e["title"].(string)
			status, _ := e["status"].(string)
			sb.WriteString(fmt.Sprintf("| %s | %s | %s |\n", specID, title, status))
		}
	}

	if len(active) > 0 {
		sb.WriteString("### Active Specs\n\n")
		writeTable(active)
		sb.WriteString("\n")
	}

	if len(archived) > 0 {
		sb.WriteString("### Archived Specs\n\n")
		writeTable(archived)
		sb.WriteString("\n")
	}

	return sb.String()
}

// formatAssessmentBlock formats an Assessment into a human-readable block
// for inclusion in prompt templates.
func formatAssessmentBlock(a Assessment) string {
	var sb strings.Builder
	sb.WriteString(fmt.Sprintf("Quality: %s\n", a.Quality))
	sb.WriteString(fmt.Sprintf("Summary: %s\n", a.Summary))
	if len(a.Gaps) > 0 {
		sb.WriteString("Gaps:\n")
		for _, gap := range a.Gaps {
			sb.WriteString(fmt.Sprintf("- %s\n", gap))
		}
	}
	if len(a.Questions) > 0 {
		sb.WriteString("Questions:\n")
		for _, q := range a.Questions {
			qText, _ := q["question"].(string)
			if qText == "" {
				qText, _ = q["text"].(string)
			}
			sb.WriteString(fmt.Sprintf("- %s\n", qText))
		}
	}
	return sb.String()
}

// formatQABlock formats question-answer pairs into a human-readable block.
// Keys are sorted lexicographically to produce deterministic output.
func formatQABlock(answers map[string]string) string {
	if len(answers) == 0 {
		return "No questions and answers provided."
	}
	keys := make([]string, 0, len(answers))
	for q := range answers {
		keys = append(keys, q)
	}
	sort.Strings(keys)
	var sb strings.Builder
	for _, q := range keys {
		sb.WriteString(fmt.Sprintf("Q: %s\nA: %s\n\n", q, answers[q]))
	}
	return sb.String()
}

// AssessmentSystemPrompt loads the assessment_system template via LoadPrompt
// and returns its content.
func AssessmentSystemPrompt(projectDir string) (string, error) {
	return LoadPrompt("assessment_system", projectDir)
}

// AssessmentUserPrompt loads and renders the assessment_user template via
// LoadPromptTemplate, substituting $spec_name, $spec_landscape_block, and $prd_text.
func AssessmentUserPrompt(prdText, specName, projectDir string, specLandscape []map[string]any) (string, error) {
	landscapeBlock := FormatSpecLandscape(specLandscape)

	return LoadPromptTemplate("assessment_user", projectDir, map[string]string{
		"spec_name":            specName,
		"prd_text":             prdText,
		"spec_landscape_block": landscapeBlock,
	})
}

// RefinementSystemPrompt loads the refinement_system template via LoadPrompt
// and returns its content.
func RefinementSystemPrompt(projectDir string) (string, error) {
	return LoadPrompt("refinement_system", projectDir)
}

// RefinementUserPrompt loads and renders the refinement_user template via
// LoadPromptTemplate, substituting $prd_text, $assessment_block, $qa_block,
// and $spec_landscape_block.
func RefinementUserPrompt(prdText string, answers map[string]string, prevAssessment Assessment, projectDir string, specLandscape []map[string]any) (string, error) {
	assessmentBlock := formatAssessmentBlock(prevAssessment)
	qaBlock := formatQABlock(answers)
	landscapeBlock := FormatSpecLandscape(specLandscape)

	return LoadPromptTemplate("refinement_user", projectDir, map[string]string{
		"prd_text":             prdText,
		"assessment_block":     assessmentBlock,
		"qa_block":             qaBlock,
		"spec_landscape_block": landscapeBlock,
	})
}

// GenerationSystemPrompt loads the generation_system template via LoadPrompt
// and returns its content.
func GenerationSystemPrompt(projectDir string) (string, error) {
	return LoadPrompt("generation_system", projectDir)
}

// GenerationUserPrompt loads generation_user_base and the artifact-specific
// template (generation_user_requirements, generation_user_test_spec, or
// generation_user_tasks) via LoadPromptTemplate, composes them with the
// appropriate variables, and returns the combined prompt.
//
// Returns an error if artifactName is not one of requirements, test_spec, or tasks.
func GenerationUserPrompt(prdText, artifactName, specID, projectDir string, priorArtifacts map[string]any, dependentInterfaces []map[string]any, specLandscape []map[string]any) (string, error) {
	// Validate artifact name.
	validArtifacts := map[string]string{
		"requirements": "generation_user_requirements",
		"test_spec":    "generation_user_test_spec",
		"tasks":        "generation_user_tasks",
	}
	templateName, ok := validArtifacts[artifactName]
	if !ok {
		return "", fmt.Errorf("GenerationUserPrompt: unrecognized artifact name %q", artifactName)
	}

	// Format blocks.
	landscapeBlock := FormatSpecLandscape(specLandscape)

	var priorArtifactsBlock string
	if len(priorArtifacts) > 0 {
		data, err := json.MarshalIndent(priorArtifacts, "", "  ")
		if err == nil {
			priorArtifactsBlock = "Previously generated artifacts:\n" + string(data)
		}
	}

	var dependentInterfacesBlock string
	if len(dependentInterfaces) > 0 {
		data, err := json.MarshalIndent(dependentInterfaces, "", "  ")
		if err == nil {
			dependentInterfacesBlock = "Dependent interfaces:\n" + string(data)
		}
	}

	// Detect project language for language_block.
	lang, hints := DetectProjectLanguage(projectDir)
	var languageBlock string
	if lang != "" {
		languageBlock = fmt.Sprintf("Project language: %s\nTooling: %s", lang, hints)
	}

	// Load the base template with variable substitution.
	basePrompt, err := LoadPromptTemplate("generation_user_base", projectDir, map[string]string{
		"artifact_name":              artifactName,
		"spec_id":                    specID,
		"prd_text":                   prdText,
		"spec_landscape_block":       landscapeBlock,
		"dependent_interfaces_block": dependentInterfacesBlock,
		"prior_artifacts_block":      priorArtifactsBlock,
		"language_block":             languageBlock,
	})
	if err != nil {
		return "", fmt.Errorf("GenerationUserPrompt: loading base template: %w", err)
	}

	// Load the artifact-specific template.
	artifactPrompt, err := LoadPrompt(templateName, projectDir)
	if err != nil {
		return "", fmt.Errorf("GenerationUserPrompt: loading artifact template: %w", err)
	}

	return basePrompt + "\n\n" + artifactPrompt, nil
}

// RepairUserPrompt loads the repair_user template via LoadPromptTemplate,
// substituting the artifact name, original content, and validation errors.
func RepairUserPrompt(artifactName string, originalContent any, validationErrors []string, projectDir string) (string, error) {
	errorsBlock := strings.Join(validationErrors, "\n")

	var contentStr string
	if originalContent != nil {
		data, err := json.MarshalIndent(originalContent, "", "  ")
		if err != nil {
			contentStr = fmt.Sprintf("%v", originalContent)
		} else {
			contentStr = string(data)
		}
	}

	return LoadPromptTemplate("repair_user", projectDir, map[string]string{
		"artifact_name":    artifactName,
		"errors_block":     errorsBlock,
		"original_content": contentStr,
	})
}
